providers/custom_provider.go

```diff
@@ -0,0 +1,126 @@
+package providers
+
+import (
+	"archive/tar"
+	"compress/gzip"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"path/filepath"
+
+	"github.com/chanzuckerberg/fogg/util"
+	"github.com/pkg/errors"
+	log "github.com/sirupsen/logrus"
+	"github.com/spf13/afero"
+)
+
+// TypeProviderFormat is the provider format such as binary, zip, tar
+type TypeProviderFormat string
+
+const (
+	// TypeProviderFormatTar is a tar archived provider
+	TypeProviderFormatTar TypeProviderFormat = "tar"
+)
+
+// CustomProvider is a custom terraform provider
+type CustomProvider struct {
+	URL    string             `json:"url,omitempty" validate:"required"`
+	Format TypeProviderFormat `json:"format,omitempty" validate:"required"`
+}
+
+// Install installs the custom provider
+func (cp *CustomProvider) Install(providerName string, dest afero.Fs) error {
+	if cp == nil {
+		return errors.New("nil CustomProvider")
+	}
+	if dest == nil {
+		return errors.New("nil fs")
+	}
+
+	tmpPath, err := cp.fetch(providerName)
+	defer os.Remove(tmpPath) // clean up
+	if err != nil {
+		return err
+	}
+	return cp.process(providerName, tmpPath, dest)
+}
+
+// fetch fetches the custom provider at URL
+func (cp *CustomProvider) fetch(providerName string) (string, error) {
+	tmpFile, err := ioutil.TempFile("", providerName)
+	if err != nil {
+		return "", errors.Wrap(err, "could not create temporary directory")
+	}
+	resp, err := http.Get(cp.URL)
+	if err != nil {
+		return "", errors.Wrapf(err, "could not get %s", cp.URL)
+	}
+	defer resp.Body.Close()
+	_, err = io.Copy(tmpFile, resp.Body)
+	return tmpFile.Name(), errors.Wrap(err, "could not download file")
```

Comment:
I presume errors.Wrap will do the right thing if err is nil?

Reply:
```
// Wrap returns an error annotating err with a stack trace
// at the point Wrap is called, and the supplied message.
// If err is nil, Wrap returns nil.
func Wrap(err error, message string) error {
	if err == nil {
		return nil
	}
```